support new webgl backend.

GreyZzzzzzXh opened the pull request.

GreyZzzzzzXh commented on examples/image_classification_tflite/ImageClassificationModel.js:
```diff
@@ -20,16 +20,14 @@ class ImageClassificationModel {
       }
       this._nn = nnNative;
-    } else if (this._backend === 'WASM' || this._backend === 'WebGL2') {
+    } else if (this._backend === 'WASM' || this._backend === 'WebGL') {
```
I renamed all **_WebGL2_** to _**WebGL**_ in polyfill and examples.

GreyZzzzzzXh commented on examples/image_classification_tflite/ImageClassificationModel.js:
```diff
@@ -20,16 +20,14 @@ class ImageClassificationModel {
-      options.useWebGL2 = true;
-    }
+    options.backend = this._backend;
     this._model = await this._nn.createModel(options);
```
Replaced `options.useWebGL2` with `options.backend`.
For polyfill, `options.backend` should be the string 'WASM' or 'WebGL'.

GreyZzzzzzXh commented:
I changed the backend name to 'WebGL', and revised  some interfaces.
@huningxin , @Wenzhao-Xiang, please help review, thanks!
These changes may effect test and benchmark, please take a look, @BruceDai @ibelem .

BruceDai commented:
@GreyZzzzzzXh yes, we will update test cases & benchmark test for testing WASM / WebGL polyfill backend, thanks

Wenzhao-Xiang commented:
@GreyZzzzzzXh Seems inception_v3 and squeezenet for image_classification_tflite have some issues with my machine:
squeezenet:

Label | Probability
 -- | --
admiral | 30.07%
refrigerator | 24.34%
wine bottle | 10.36%

inception_v3:

Label | Probability
-- | --
fountain | 100.00%
toilet tissue | 0.00%
bolete | 0.00%

But the result was correct in @GreyZzzzzzXh  and @pinzhenx 's machine with the same code.
Does tensorflow.js have some requests for machine?

Wenzhao-Xiang commented:
And the results have some difference with PC and mobile phone(One plus 3T, chrome) with WebGL backend. For example, with mobilenet_v1:
PC:

Label | Probability
-- | --
bee eater | 95.59%
partridge | 1.73%
indigo bunting | 1.71%

Mobile phone:

Label | Probability
-- | --
bee eater | 95.41%
partridge | 1.69%
indigo bunting | 1.67%

It seems like the same issue as what Wenyao write in his graduation paper: 

> On mobile devices, WebGL uses 16-bit signed integers to store data by default, while on the computer,
> WebGL uses 32-bit signed integers to store data by default.

GreyZzzzzzXh commented:
@Wenzhao-Xiang, thanks for testing. There are still some issues on mobile devices.

> On mobile devices, WebGL uses 16-bit signed integers to store data by default, while on the computer,
WebGL uses 32-bit signed integers to store data by default.

May be like this, but if  so we have to revise the source code of tfjs.

It may take some time to figure out, so.. maybe we should still remain _WebGL2_ backend and name this new backend _TFJS_ temporarily.

huningxin commented:
> It may take some time to figure out, so.. maybe we should still remain _WebGL2_ backend and name this new backend _TFJS_ temporarily.

Can we make it a reproducible test case for TensorFlow.js team? For example, put it to a public accessible link in your github pages. We can file a bug to TF.js. And I can help bridge TF.js folks to look into it.

GreyZzzzzzXh commented:
I host the modified code on my github pages. 
Visit https://greyzzzzzzxh.github.io/webml-examples to see the results.

huningxin commented:
> I host the modified code on my github pages.
> Visit https://greyzzzzzzxh.github.io/webml-examples to see the results.

Thanks for doing that!

Besides the examples link, could you please give the steps to reproduce the issue? For example, which example, which backend, which model? And please give the expected results. And what is the incorrect result when the issue happens? Also please give the test configuration, e.g. Chrome version (got from chrome://version/), GPU driver version (got from chrome://gpu/) etc.,. Please consult @BruceDai for this kind of bug reporting. Thanks!

GreyZzzzzzXh commented:
@huningxin , ok i'll do more test and provide detail info next week, thanks!

GreyZzzzzzXh commented:
**Test Env:**
Chrome Version: 70.0.3538.80 (official build) (32-bit)
Platform: Android 8.1.0, Pixel 2
GPU driver version: 258.0

**Expected Result:**
The result should be the same as that on the computer side.

Mobilenet v1:

| # | Label | Probability |
-- | -- | -- 
1 | bee eater | 95.59%
2 | jacamar | 1.73%
3 | brambling | 1.71%

Mobilenet v2:

| # | Label | Probability |
-- | -- | --
1 | bee eater | 84.14%
2 | indigo bunting | 1.07%
3 | brambling | 0.76%

Inception v3:

| # | Label | Probability |
-- | -- | --
1 | bee eater | 96.31%
2 | partridge | 0.11%
3 | indigo bunting | 0.04%

Squeezenet:

| # | Label | Probability |
-- | -- | --
1 | bee eater | 96.71%
2 | goldfinch | 1.77%
3 | ladybug | 0.45%


**Actual Result:**

The output doesn't match the expected result.

For example:
Mobilenet v1:

| # | Label | Probability
-- | -- | --
1 | bee eater | 95.41%
2 | brambling | 1.69%
3 | jacamar | 1.67%

Same issues on other models.

**How to Reproduce:**

- Launch chrome and visit https://greyzzzzzzxh.github.io/webml
- Choose _Classification Image Demo_
- Change backend to _WebGL_
- Change different models

GreyZzzzzzXh commented:
Results and reproduction steps are described above, please take a look @huningxin .

huningxin commented:
@GreyZzzzzzXh , thanks! How about the expected results? And please list the devices which can deliver expected results. That would be also helpful.

GreyZzzzzzXh commented:
> How about the expected results

data in the first four tables is expected result.


> the devices which can deliver expected results

PC platform, e.g.

- Linux ubuntu 16.04, Chrome 69.0.3497.100 (Official Build) (64-bit)
- Windows 10, Chrome 70.0.3538.102 (Official Build) (32-bit)

GreyZzzzzzXh commented:
Besides,  visit https://greyzzzzzzxh.github.io/webml/test/index.html?backend=webgl&grep=CTS for case testing.

computer side: 

> passes: 127
> failures: 6

but many cases fail on mobile device:

> passes: 71
> failures: 62

huningxin commented:
@GreyZzzzzzXh , please complete the actual results in https://github.com/intel/webml-polyfill/pull/297#issuecomment-438516825. Thanks!

BTW, when testing on the device with incorrect results, are there any errors reported in console?

GreyZzzzzzXh commented:
> please complete the actual results in [#297 (comment)](https://github.com/intel/webml-polyfill/pull/297#issuecomment-438516825)

done.


> when testing on the device with incorrect results, are there any errors reported in console?

 no errors or warnings reported in console.

huningxin commented:
Thanks for these details!

huningxin commented:
@GreyZzzzzzXh , could you please take a look at https://github.com/tensorflow/tfjs/issues/265. It sounds like tfjs uses float16 on mobile. Is that the root cause?

GreyZzzzzzXh commented:
This precision issue can be fixed by upgrading GLSL version to _300 es_.
WebGL1.0 dont work now so we just use webgl2.0 backend and still call it _WebGL2_.

huningxin commented:
> This precision issue can be fixed by upgrading GLSL version to _300 es_.

Great finding! Could you tell how to change it? Can we make it configurable?

GreyZzzzzzXh commented:
@huningxin , yes i'll record the changes in `README.md`, and now i import modified tfjs-core in `src/nn/webgl2/tfjs-core` for use.

@BruceDai , for test, just build and run https://github.com/GreyZzzzzzXh/webml-polyfill/tree/webgl_backend as usual, thx.

huningxin commented:
> yes i'll record the changes in `README.md`, and now i import modified tfjs-core in `src/nn/webgl2/tfjs-core` for use.

According to https://github.com/GreyZzzzzzXh/tfjs-core/commit/39a56bf75803ae9327bbf160189f0d08f8a06416, could you please explain the root cause and your method of "upgrade GLSL to version 300 es"?

I found you check-in tfjs-core, I would suggest to avoid that. Let's figure out what need to be fixed in tfjs-core. Then propose a fix to tfjs-core repo.

GreyZzzzzzXh commented:
> the root cause

In general, webgl1.0 and 2.0 will use different versions of the shading language (_GLGL100_ for webgl1.0 and _GLSL300es_ for webgl2.0). But in tfjs, only _GLSL100_ is used as the shading language for webgl1.0 and 2.0. 

The inaccuracy on the phone seems to be because the version of GLSL also has an impact on accuracy. After I changed the _GLSL100_ to _300es_, the float can achieve 32-bit accuracy on the phone, and originally only 16 bits (tested with `tf.ENV.backend.floatPrecision()`).


> your method of "upgrade GLSL to version 300 es"

I added some comments about how to upgrade GLSL version in [GreyZzzzzzXh/tfjs-core@39a56bf](https://github.com/GreyZzzzzzXh/tfjs-core/commit/39a56bf75803ae9327bbf160189f0d08f8a06416).
There are five main places that need to be changed:
1. Declare the shading language version in the shader code as `#version 300 es`.
2. Replace **_attribute_** with **_in_**.
3. Replace **_varying_**  with **_in/out_**.
4. Replace **_texture2D_** with **_texture_**.
5. There is no built-in variable _**gl_FragColor**_ in _GLSL300es_, so we need to define an **_out_** variable for the output.

Besides, i set `precision highp sampler2D;` and made some changes related to function `round()`.
See [GreyZzzzzzXh/tfjs-core@39a56bf](https://github.com/GreyZzzzzzXh/tfjs-core/commit/39a56bf75803ae9327bbf160189f0d08f8a06416) for more detail.

GreyZzzzzzXh commented:
> Let's figure out what need to be fixed in tfjs-core. Then propose a fix to tfjs-core repo.

Yeah, the best way is to solve this problem by tfjs team. But if it takes a while to fix, I suggest to import the modified tfjs for temporary use.

huningxin commented:
> Yeah, the best way is to solve this problem by tfjs team

Please go ahead to file a bug and open a PR to tfjs-core with your solution.

> I suggest to import the modified tfjs for temporary use.

Please don't import source code. If we want to maintain a version of tfjs-core, you can publish your version in npm and npm install from there.



> originally only 16 bits (tested with `tf.ENV.backend.floatPrecision()`).

Probably, we can report this out. Then test cases can handle the lower precision backend differently. @BruceDai

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+  prepareModel() {
+    this._model._operands.forEach(operand => {
+      if (utils.isTensor(operand.type)) {
+        let type = this._getOperandType(operand.type);
```
`const`?

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+      output.buffer.set(operand.dataSync());
+    });
+    
+    // console.log(tf.memory());
```
remove this comment?

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+    }
+
+    inputs.forEach(input => {
+      let operand = this._operands[input.index];
```
const?

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+
+    inputs.forEach(input => {
+      let operand = this._operands[input.index];
+      let inputTensor = tf.tensor(input.buffer, operand.shape, operand.dtype);
```
const?

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+
+    switch(op) {
+      case OperationCode.ADD: {
+        let in1 = operands[inputs[0]];
```
const?

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+        output.assign(activation(tf.add(in1, in2)));
+      } break;
+      case OperationCode.MUL: {
+        let in1 = operands[inputs[0]];
```
const?

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+          throw new Error('Invalid parameters number of CONV_2D');
+        }
+        let i = 0;
+        let input = operands[inputs[i++]];
```
const?

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+        output.assign(activation(tf.mul(in1, in2)));
+      } break;
+      case OperationCode.CONV_2D: {
+        let inCount = inputs.length;
```
const?

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+          strideH = operands[inputs[i++]].value[0];
+          activation = FuseFunctionMap.get(operands[inputs[i++]].value[0]);
+          output.assign(activation(
+            input.conv2d(filter, [strideH, strideW], padding).add(bias)));
```
4-spaces indent for the continuous line.

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+          strideH = operands[inputs[i++]].value[0];
+          activation = FuseFunctionMap.get(operands[inputs[i++]].value[0]);
+          output.assign(activation(
+            input.pad([[0, 0], [paddingTop, paddingBottom], [paddingLeft, paddingRight], [0, 0]])
```
ditto

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+          output.assign(activation(
+            input.conv2d(filter, [strideH, strideW], padding).add(bias)));
+        } else {
+          let paddingLeft = operands[inputs[i++]].value[0];
```
const

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+        }
+      } break;
+      case OperationCode.DEPTHWISE_CONV_2D: {
+        let inCount = inputs.length;
```
const

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+          depthMultipler = operands[inputs[i++]].value[0];
+          activation = FuseFunctionMap.get(operands[inputs[i++]].value[0]);
+          output.assign(activation(
+            input.pad([[0, 0], [paddingTop, paddingBottom], [paddingLeft, paddingRight], [0, 0]])
```
4-spaces indent

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+      } break;
+      case OperationCode.AVERAGE_POOL_2D:
+      case OperationCode.MAX_POOL_2D: {
+        let inCount = inputs.length;
```
const

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+        }
+      } break;
+      case OperationCode.SOFTMAX: {
+        let input = operands[inputs[0]];
```
const

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+        output.assign(input.mul(tf.scalar(beta)).softmax());
+      } break;
+      case OperationCode.RESHAPE: {
+        let input = operands[inputs[0]];
```
ditto

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+        let input = operands[inputs[0]];
+        let targetShape = operands[inputs[1]];
+        let output = operands[outputs[0]];
+        output.assign(input.reshape(targetShape.dataSync()));
```
do we need to do `dataSync`? I understand it leads to memory read back from GPU to CPU which is bad for performance.

GreyZzzzzzXh replied:
According to the definition of [reshape in NN API](https://github.com/intel/webml-polyfill/blob/master/src/nn/Enums.js#L952), targetShape is a 1-D tensor whose value is stored in GPU. But for [tf.reshape](https://js.tensorflow.org/api/0.13.3/#reshape), target shape should be an array of integers..

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+        output.assign(input.reshape(targetShape.dataSync()));
+      } break;
+      case OperationCode.CONCATENATION: {
+        if (outputs.length < 1 || inputs.length < 2) {
```
simplify the condition?

GreyZzzzzzXh replied:
since tfjs will give detailed reasons for errors, can we remove this check?

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+        let bias = operands[inputs[2]];
+        let activation = FuseFunctionMap.get(operands[inputs[3]].value[0]);
+        let output = operands[outputs[0]];
+        let batchSize = input.shape[0];
```
`batchSize` should be `product (input.shape) / weights.shape[1]`, according to android nn API spec: https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0aaada7a3dbaf4676aba560c933ff610c5

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+      let op = operation.type;
+      switch(op) {
+        case OperationCode.CONV_2D: {
+          // NHWC -> HWCN
```
Please add link in the document about tf.js conv weights format.

huningxin commented on src/nn/webgl/WebGLModel.js:
```diff
@@ -0,0 +1,328 @@
+          filter.dispose();
+        } break;
+        case OperationCode.DEPTHWISE_CONV_2D: {
+          // [1, filterH, filterW, outChannels] -> [filterH, filterW, inChannels, depthMultipler]
```
ditto.

huningxin commented:
@GreyZzzzzzXh , finish my review with some comments. Please take a look. Thanks!

GreyZzzzzzXh commented:
I made some changes, PTAL, thanks! @huningxin

huningxin commented:
The Travis CI fails due to "chrome installation error". @ibelem , could you or someone please have a check?

ibelem commented:
@huningxin Please feel free to merge this PR since there is the issue of Travis CI and passed with AppVeyor. We are reported Travis CI issue to [upstream](https://travis-ci.community/t/macos-brew-install-google-chrome-fail/1290) and also trying other workrounds.

huningxin commented:
Thanks for the great work. Looks good to me!
